iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/ClusterTopology.java

```diff
@@ -0,0 +1,167 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.queryengine.plan;
+
+import org.apache.iotdb.common.rpc.thrift.TConsensusGroupId;
+import org.apache.iotdb.common.rpc.thrift.TDataNodeLocation;
+import org.apache.iotdb.common.rpc.thrift.TRegionReplicaSet;
+import org.apache.iotdb.db.conf.IoTDBDescriptor;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ClusterTopology {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ClusterTopology.class);
+  private final Integer myself;
+  private volatile Map<Integer, TDataNodeLocation> dataNodes;
+  private volatile Map<Integer, Set<Integer>> topologyMap;
+  private volatile boolean isPartitioned;
+
+  public static ClusterTopology getInstance() {
+    return ClusterTopologyHolder.INSTANCE;
+  }
+
+  public TRegionReplicaSet getReachableSet(TRegionReplicaSet origin) {
+    if (!isPartitioned || origin == null) {
+      return origin;
+    }
+    final Set<Integer> reachableToMyself = Collections.unmodifiableSet(topologyMap.get(myself));
+    final List<TDataNodeLocation> locations = new ArrayList<>();
+    for (final TDataNodeLocation location : origin.getDataNodeLocations()) {
+      if (reachableToMyself.contains(location.getDataNodeId())) {
+        locations.add(location);
+      }
+    }
+    return new TRegionReplicaSet(origin.getRegionId(), locations);
+  }
+
+  public <T> Set<Map.Entry<TRegionReplicaSet, T>> filterReachableCandidates(
+      Set<Map.Entry<TRegionReplicaSet, T>> input) {
+    final List<TRegionReplicaSet> allSets =
+        input.stream().map(Map.Entry::getKey).collect(Collectors.toList());
+    final List<TRegionReplicaSet> candidates = getReachableCandidates(allSets);
+    final Map<TConsensusGroupId, TRegionReplicaSet> newSet = new HashMap<>();
+    candidates.forEach(set -> newSet.put(set.getRegionId(), set));
+    final Map<TRegionReplicaSet, T> candidateMap = new HashMap<>();
+    for (final Map.Entry<TRegionReplicaSet, T> entry : input) {
+      final TConsensusGroupId gid = entry.getKey().getRegionId();
+      if (newSet.containsKey(gid)) {
+        candidateMap.put(newSet.get(gid), entry.getValue());
+      }
```

Comment:
directly call newSet.get(gid) and use value != null to avoid calling containsKey

Reply:
Good advice! Fixed.